userCode/awqms/dag.py

```diff
@@ -0,0 +1,208 @@
+            LOGGER.warning(f"Station {station['@iot.id']} already exists")
+            return
+
+    resp = requests.post(url_join(API_BACKEND_URL, "Things"), json=station)
+    if not resp.ok:
+        LOGGER.error(f"Failed posting thing: {station}")
+        raise RuntimeError(resp.text)
+
+    return
+
+
+@asset(partitions_def=station_partition, group_name="awqms")
+def awqms_datastreams(awqms_metadata: StationData) -> list[Datastream]:
+    thingid = awqms_metadata.MonitoringLocationId
+
+    datastreams: list[Datastream] = []
+    for datastream in awqms_metadata.Datastreams:
+        if datastream.observed_property not in POTENTIAL_DATASTREAMS:
+            continue
+
+        datastreams.append(
+            to_sensorthings_datastream(
+                awqms_metadata,
+                POTENTIAL_DATASTREAMS[datastream.observed_property],
+                datastream.observed_property,
+                thingid,
+            )
+        )
+
+    assert len(datastreams) > 0, f"No datastreams found for {thingid}"
+
+    return datastreams
+
+
+@asset(partitions_def=station_partition, deps=[post_awqms_station], group_name="awqms")
+def post_awqms_datastreams(awqms_datastreams: list[Datastream]):
+    # check if the datastreams exist
+    for datastream in awqms_datastreams:
+        url = url_join(API_BACKEND_URL, f"Datastreams('{datastream.iotid}')")
+        resp = requests.get(url)
+        if resp.status_code == 404:
+            msg = f"Datastream {datastream.iotid} not found. Posting..."
+            LOGGER.info(msg)
+        elif not resp.ok:
+            msg = f"Failed checking if datastream '{datastream.iotid}' exists"
+            LOGGER.error(msg)
+            raise RuntimeError(resp.text)
+        else:
+            id = resp.json()["@iot.id"]
+            if id == datastream.iotid:
+                LOGGER.warning(f"Datastream {datastream.iotid} already exists")
+                continue
+
+        resp = requests.post(
+            url_join(API_BACKEND_URL, "Datastreams"),
+            json=datastream.model_dump(by_alias=True),
+        )
+        if not resp.ok:
+            LOGGER.error(f"Failed posting datastream: {datastream}")
+            raise RuntimeError(resp.text)
+
+
+@asset(
+    partitions_def=station_partition, deps=[post_awqms_datastreams], group_name="awqms"
+)
+async def awqms_observations(
+    awqms_metadata: StationData, awqms_datastreams: list[Datastream]
+) -> bool:
+    associatedThing = awqms_metadata.MonitoringLocationId
+
+    observations: dict[int, Observation] = {}
+
+    async def fetch_and_process(datastream: Datastream):
+        observations_ids = fetch_observation_ids(datastream.iotid)
+        LOGGER.info(f"Fetching observations for {datastream.iotid}")
+        for result in fetch_observations(datastream.description, associatedThing):
+            if not result["ResultValue"]:
+                continue
+
+            id = "".join([datastream.iotid, result["StartDateTime"]])
```

Comment:
I may be misunderstanding something here, but isn't it more idiomatic to just concat the two together instead of join which I would expect to be used on a long iterable?

```suggestion
            id = f"{datastream.iotid}{result["StartDateTime"}"
```

Reply:
Had included additional properties in the hash - with two I agree